### ISSUE TITLE
Add option in UpdateXXXInput to select if Update should trigger Reactives

### ISSUE THREAD

**Issue by JulianBo** [rstudio/shiny#2324]
When having a reactive that a) depends on two inputs and b) Updates one of these inputs, the reactive will calculate twice. To avoid this, an option in updateXXXInput would be nice to select if update should (not) trigger reactives. 

See: https://stackoverflow.com/questions/54560439/shiny-update-input-without-reactives-getting-triggered

**Comment by r2evans** [comment]
Another way to approach this is that `isolate(updateSliderInput(...))` did not isolate the reactivity of the slider input. The code from the linked SO question is:

```r
library(shiny)

ui <- fluidPage(
   titlePanel("Update Slider - Isolate reaction?"),

   sidebarLayout(
      sidebarPanel(
         sliderInput("bins",
                     "Number of bins:",
                     min = 1,
                     max = 50,
                     value = 30),
         actionButton("set1", "set slider 'bins'$value=20"),
         actionButton("set2", "'ISOLATED' set slider 'bins'$value=20 ")
      ),

      mainPanel(
         textOutput("sliderValue")
      )
   )
)

# Define server logic 
server <- function(input, output, session) {
   output$sliderValue <- renderText(input$bins) 

   observeEvent(input$set1,{
     updateSliderInput(session,"bins",value=20)
   })
   observeEvent(input$set2,{
     ## Is there any possibility to update the slider without 'sliderValue' changing?
     #isolate does not work
     isolate(
       updateSliderInput(session,"bins",value=20 )
       )
   })
}

shinyApp(ui = ui, server = server)
```

**Comment by daattali** [comment]
I've ran into cases in the past when I also wished that something like this existed. But I haven't thought about it much longer than "huh, I wish I could do that."

I do wonder what exactly this would look like though -- if you don't want it to trigger ANY reaction, then what does it really mean? What actually happened when the value changed? If you ask for the value of input$bins later on, should it not be the new updated value? And if it should, then at what point does it become this new value? If it doesn't return the new value right now, then the app will be in an inconsistent/false state. 

Perhaps I'm overcomplicating this, just a thought I just had.

**Comment by r2evans** [comment]
I'd think the point of "update without triggering reactivity" would be just *"change the value as it is stored and shown in the HTML object, nothing more"*. Anything that explicitly looked for the value (perhaps due to a different reactive trigger) would get the updated value.


- *"if you don't want it to trigger ANY reaction, then what does it really mean?"* Literally just update the value in the HTML object. If a user looks at the (say) slider, it will show the new number. If your meaning of reactivity means "update the HTML representation", then that should certainly be done ... my interpretation was not going that deep. But nothing more.

- *"What actually happened when the value changed?"* Other than updating the HTML object UI (moving the slider thingy from one value to another), nothing. If there are constraints that rely on that (e.g., the slider is indicating an upper-bound on something else), then tough. As the programmer, if you are explicitly putting the "system" in an inconsistent state, then you bought all ramifications of it.

- *"If you ask for the value of input$bins later on, should it not be the new updated value?"* Yes, it should absolutely be the new (not-yet-propagated) value. There should be no "delay" in this, partly because that defies the intuitive logic of "update the value, with or without follow-on reactivity", but also because I know that adds a lot of complexity. More the first. That's not to say that a genius programmer can't affect a delayed-no-reactive-update, but again that's a special corner case. I'd think this "no-reactive-update" would need to be as orthogonal and simple as possible.

- *"the app will be in an inconsistent state"*, absolutely yes. But that can already happen with explicit programmer action. What I'm suggesting here falls in line with that: it must be an intentional action by the programmer to put things in this state. It is already feasible with (say) "press [this button] to update &lt;that plot&gt;", and may be a desirable design choice.

**Comment by daattali** [comment]
If the goal is to merely update the value shown to the user, just on the surface level in the HTML, then that's probably not too difficult to achieve right now with a little bit of javascript. That would not have sufficed me for my usecases though, that's why I was asking all these questions!

**Comment by JulianBo** [comment]
I agree with most of r2evans answers, except

·         “"If you ask for the value of input$bins later on, should it not be the new updated value?" Yes, it should absolutely be the new (not-yet-propagated) value. There should be no "delay" in this, partly because that defies the intuitive logic of "update the value, with or without follow-on reactivity"”

and

·         “the app will be in an inconsistent state", absolutely yes.”

I think the value of input$bins delivered to the programme should still be the old one. And that is why the app itself will not be in an inconsistent state – there will only be an inconsistency between the values shown to the user and the internal values, but only until the user updates the value manually, and this new value triggers all reactives as normal. 

Basically, I think of this as temporarily switching the inputs’ function to an output, to show some calculated state to the user. 

This suffices for my use-cases, and I was hoping this could be done by some simple javascript.  

 

 

Von: Dean Attali [mailto:notifications@github.com] 
Gesendet: Montag, 25. Februar 2019 19:56
An: rstudio/shiny
Cc: Julian Bothe; Author
Betreff: Re: [rstudio/shiny] Add option in UpdateXXXInput to select if Update should trigger Reactives (#2324)

 

If the goal is to merely update the value shown to the user, just on the surface level in the HTML, then that's probably not too difficult to achieve right now with a little bit of javascript. That would not have sufficed me for my usecases though, that's why I was asking all these questions!

**Comment by daattali** [comment]
@JulianBo sounds like what you want can indeed by solved with a short javascript chunk. But I wouldn't recommend that, the value that the user will see is not going to be true, I think that's a bad idea.

Here's a very simple example to demonstrate why (along with a very simplistic way to implement what you asked for):

```
library(shiny)

ui <- fluidPage(
  shinyjs::useShinyjs(),
  actionButton("go", "Set num to 10 ONLY for the client"),
  numericInput("num", "Number", 5),
  "Double:", textOutput("double", inline = TRUE)
)

server <- function(input, output, session) {
  output$double <- renderText({
    input$num * 2
  })
  
  observeEvent(input$go, {
    shinyjs::runjs('$("#num").val(10)')
  })
}

shinyApp(ui, server)
```

When you update the numeric input to show 10 to the show, now the output that is supposed to show the number doubles is no longer correct. I don't recommend this.

**Comment by r2evans** [comment]
I agree with @daattali , if you're going to update the value sans reactivity, I can't imagine not showing it in the UI. That to me would seem doubly-inconsistent in that the value being shown to the user is neither what is stored internally nor what any of the tables/plots were (at some point) using. (The concept of "inconsistent state" could possibly be conceived as a little deceitful, but this double-inconsistency seems even worse.)
